File: JavaOpenDataAccess/src/OpenDataAccess/Data/Query.java
package OpenDataAccess.Data;

import java.util.List;

public class Query implements IQuery {
    public String GetTableName() {
        return _TableName;
    }

    public void SetTableName(String value) {
        _TableName = value;
    }

    public String GetWithSql() {
        return _WithSql;
    }

    public void SetWithSql(String value) {
        _WithSql = value;
    }

    public String GetEntityName() {
        return _EntityName;
    }

    public void SetEntityName(String value) {
        _EntityName = value;
    }

    public IDataParameterList GetParameterList() {
        return _ParameterList2;
    }

    public void SetParameterList(IDataParameterList value) {
        _ParameterList2 = value;
    }

    private String _WithSql = "";
    private String _TableName = "";
    private String _EntityName = "";
    private String _JoinSql = "";
    private String _WhereSql = "";
    private String _OrderBySql = "";
    private String _GroupBySql = "";
    private String _FieldSql = "*";
    private String _Sql = "";
    private IDataParameterList _ParameterList = null;
    private IDataParameterList _ParameterList2 = null;

    public Query(String tableName) {
        this._TableName = tableName;
    }

    public Query(String tableName, String entityName) {
        this._TableName = tableName;
        this._EntityName = entityName;
    }

    public IQuery Join(List<JoinStatement> joinList) {
        _JoinSql = Common.GetJoinStatement(joinList);
        return this;
    }

    public IQuery Join(String joinSql) {
        _JoinSql = joinSql;
        return this;
    }

    public IQuery Where(String whereSql, IDataParameterList parameterList) {
        _WhereSql = " ".concat(whereSql).concat(" ");
        _ParameterList = parameterList;
        return this;
    }

    public IQuery Where(List<WhereStatement> whereList, IDataParameterList parameterList) {
        _WhereSql = Common.GetWhereStatement(whereList);
        _ParameterList = parameterList;
        return this;
    }

    public IQuery Select(String fieldSql) {
        _FieldSql = " ".concat(OpenDataAccess.Utility.Common.IsNullOrEmpty(fieldSql) ? "*" : fieldSql).concat(" ");
        return this;
    }

    public IQuery Select(List<String> fieldList) {
        _FieldSql = " ".concat(Common.GetFields(fieldList)).concat(" ");
        return this;
    }

    public IQuery OrderBy(String orderBySql) {
        _OrderBySql = " ".concat(orderBySql).concat(" ");
        return this;
    }

    public IQuery OrderBy(List<OrderByStatement> orderByList) {
        _OrderBySql = " ".concat(Common.GetOrderByStatement(orderByList)).concat(" ");
        return this;
    }

    public IQuery GroupBy(List<String> groupByList) {
        _GroupBySql = " ".concat(Common.GetGroupByFields(groupByList)).concat(" ");
        return this;
    }

    public IQuery GroupBy(String groupBySql) {
        _GroupBySql = " ".concat(groupBySql).concat(" ");
        return this;
    }

    public IQuery SetSql(String sqlText, IDataParameterList parameterList) {
        _Sql = sqlText;
        if (parameterList != null) {
            _ParameterList = parameterList;
        }
        return this;
    }

    public String ToSql() {
        String sqlText = "";
        if (OpenDataAccess.Utility.Common.IsNullOrEmpty(_Sql)) {
            sqlText = String.format("%%%%%%%%", "select ", _FieldSql, " from ", _TableName, _JoinSql, _WhereSql, _GroupBySql, _OrderBySql);
        } else {
            sqlText = _Sql;
        }
        _ParameterList2 = _ParameterList;
        _FieldSql = "*";
        _GroupBySql = "";
        _JoinSql = "";
        _OrderBySql = "";
        _WhereSql = "";
        _ParameterList = null;
        _Sql = "";
        return sqlText;
    }

    public String ToWhereSql() {
        _ParameterList2 = _ParameterList;
        return _WhereSql;
    }

    public String ToGroupSql() {
        return _GroupBySql;
    }

    public String ToSelectSql() {
        return _FieldSql;
    }

    public String ToOrderBySql() {
        return _OrderBySql;
    }
}

File: OpenDataAccessCore/OpenDataAccessCore.Data/Query.cs
using System.Collections.Generic;
using System.Data;

namespace OpenDataAccessCore.Data
{
    public class Query : IQuery
    {
        public string EntityName { get; set; }
        public string TableName { get; set; }
        public string WithSql { get; set; }
        private string _JoinSql = string.Empty;
        private string _WhereSql = string.Empty;
        private string _OrderBySql = string.Empty;
        private string _GroupBySql = string.Empty;
        private string _FieldSql = "*";
        private string _Sql = string.Empty;
        private List<IDbDataParameter> _ParameterList = null;
        public List<IDbDataParameter> ParameterList { get; set; }

        public Query(string tableName)
        {
            this.TableName = tableName;
        }

        public Query(string tableName, string entityName)
        {
            this.TableName = tableName;
            this.EntityName = entityName;
        }

        public IQuery Join(List<JoinStatement> joinList)
        {
            _JoinSql = Common.GetJoinStatement(joinList);
            return this;
        }

        public IQuery Join(string joinSql)
        {
            _JoinSql = string.Concat(" ", joinSql, " "); ;
            return this;
        }

        public IQuery Where(string whereSql, List<IDbDataParameter> parameterList = null)
        {
            _WhereSql = string.Concat(" ", whereSql, " ");
            _ParameterList = parameterList;
            return this;
        }
        public IQuery Where(List<WhereStatement> whereList, List<IDbDataParameter> parameterList = null)
        {
            _WhereSql = Common.GetWhereStatement(whereList);
            _ParameterList = parameterList;
            return this;
        }

        public IQuery Select(string fieldSql)
        {
            _FieldSql = string.Concat(" ", string.IsNullOrEmpty(fieldSql) ? "*" : fieldSql, " ");
            return this;
        }

        public IQuery Select(List<string> fieldList)
        {
            _FieldSql = string.Concat(" ", Common.GetFields(fieldList), " ");
            return this;
        }

        public IQuery OrderBy(string orderBySql)
        {
            _OrderBySql = string.Concat(" ", orderBySql, " ");
            return this;
        }

        public IQuery OrderBy(List<OrderByStatement> orderByList)
        {
            _OrderBySql = string.Concat(" ", Common.GetOrderByStatement(orderByList), " ");
            return this;
        }

        public IQuery GroupBy(List<string> groupByList)
        {
            _GroupBySql = string.Concat(" ", Common.GetGroupByFields(groupByList), " ");
            return this;
        }

        public IQuery GroupBy(string groupBySql)
        {
            _GroupBySql = string.Concat(" ", groupBySql, " ");
            return this;
        }

        public IQuery SetSql(string sqlText, List<IDbDataParameter> parameterList = null)
        {
            _Sql = sqlText;
            if (parameterList != null)
            {
                _ParameterList = parameterList;
            }
            return this;
        }

        public string ToSql()
        {
            if (TableName.StartsWith("(") && TableName.EndsWith(")")) TableName += " t";
            
            string sqlText = string.Empty;
            if (string.IsNullOrEmpty(_Sql))
            {
                sqlText = string.Concat("select ", _FieldSql, " from ", TableName, _JoinSql, _WhereSql, _GroupBySql, _OrderBySql);
            }
            else
            {
                sqlText = _Sql;
            }
            if (!string.IsNullOrEmpty(WithSql)) sqlText = WithSql + " " + sqlText;
            ParameterList = _ParameterList;
            _FieldSql = "*";
            _GroupBySql = string.Empty;
            _JoinSql = string.Empty;
            _OrderBySql = string.Empty;
            _WhereSql = string.Empty;
            _ParameterList = null;
            _Sql = string.Empty;
            return sqlText;
        }

        public string ToWhereSql()
        {
            ParameterList = _ParameterList;
            return _WhereSql;
        }

        public string ToGroupSql()
        {
            return _GroupBySql;
        }

        public string ToSelectSql()
        {
            return _FieldSql;
        }

        public string ToOrderBySql()
        {
            return _OrderBySql;
        }
    }
}
